src/libstore/binary-cache-store.cc

```diff
@@ -254,6 +256,24 @@ void BinaryCacheStore::addToStore(const ValidPathInfo & info, const ref<std::str
     if (repair || !fileExists(narInfo->url)) {
         stats.narWrite++;
         upsertFile(narInfo->url, *narCompressed);
+#if ENABLE_IPFS
+        if (publishToIPFS) {
+          try {
+            auto narPath = narInfo->url;
+            auto pos = narInfo->url.find_last_of("/");
+            if (pos != std::string::npos) {
+              narPath = narInfo->url.substr(pos+1);
+            }
+            ipfsHash = IPFSAccessor::addFile(narPath, *narCompressed);
+
+            if (!ipfsHash.empty()) {
```

Comment:
The file is uploaded through the HTTP API and a lot can go wrong there.
As for the cpp part, this is the relevant code:
https://github.com/vasild/cpp-ipfs-api/blob/master/src/client.cc#L164


I have tested the code in this PR with more paths after posting this RFC and quite some requests failed silently as the function is `void` and does not `raise` anything. As this is unacceptable, the next iteration of the implementation needs to include error handling when adding files if this feature will be even included.
**Reason:**
My research into IPFS revealed that one needs to pay attention to a lot of things (it's not ftp after all). These include trivial things like selecting a `chunker` and rather complex tasks like collecting garbage after *n* `ipfs add`s while not throwing away unpinned content (race condition).
From a design perspective, adding nars/nix store paths to IPFS must be handles by a separate tool as this is to much complexity to go into Nix (following Ken Thompson's philosophy here).
(Also part of the reason for https://github.com/NixOS/nix/pull/1167#issuecomment-273938464)